pretyping/pretyping.ml

```diff
@@ -625,8 +625,17 @@ let rec pretype k0 resolve_tc (tycon : type_constraint) (env : ExtraEnv.t) evdre
         match tycon with
         | Some ty -> ty
         | None ->
-          new_type_evar env evdref loc in
-        { uj_val = e_new_evar env evdref ~src:(loc,k) ~naming ty; uj_type = ty }
+           new_type_evar env evdref loc in
+      let store =
+        if Flags.is_program_mode () then
+          match k with
+          | Evar_kinds.NamedHole _
+          | Evar_kinds.QuestionMark _
+          | Evar_kinds.ImplicitArg (_, _, false) -> Evarsolve.obligation_store
```

Comment:
I don't have a good solution here, but evar kinds are fragile in the sense that they occasionally have  a semantics for this or that, e.g. `VarInstance` and `EvarGoal` for evars in goals, or here `NamedHole`, `QuestionMark`, `ImplicitArgs` for obligations, but this semantics is implicit. For instance why `BinderType`, or `CasesType` are not obligations?